### ISSUE TITLE
'timestep_sampling=flux_shift' can offer quality improvements, but barely mentioned in docs

### ISSUE THREAD

**Issue by araleza** [kohya-ss/sd-scripts#1958]
So I just discovered this Flux (SD3 branch) parameter:
```
timestep_sampling=flux_shift
```
Previously I'd been using
```
timestep_sampling=shift
```
due to:
1) The README.md having `timestep_sampling=shift` in the example training command
2) The extensive diagrams showing the effect of changing the `discrete_flow_shift` value and the effect of different `sigmoid` values - none of which apply to the `timestep_sampling=flux_shift` mode. The space given to these modes in the README.md kind of implies (in my opinion) that these are the modes to use.
3) Almost zero casual references to `flux_shift` on this sd-scripts repository. Searching for 'flux_shift' in the GitHub search bar returns very few results.

But now I've stumbled across this mode and tried it, my sample images are immediately more realistic.

I don't think enough people are aware of this mode and how good it can be, so I'm mostly filing this issue to raise its profile. @kohya-ss , should the README.md be updated to make `flux_shift` be the default in the example training command for Flux?

@recris, we've been talking about how to avoid the vertical lines artifact, and I came across this thread where someone has a similar-looking horizontal line artifact, and claims he 'fixed' it by using `flux_shift`:

https://github.com/kohya-ss/sd-scripts/issues/1948

Worth a try if you have some line artifacts and you currently aren't using this timestep sampling mode?

**Comment by recris** [comment]
I've been using `flux_shift` for a long time, it does not fully solve the vertical lines issue.

The only thing that seems to fully avoid the problem is to train with larger batch sizes and/or low learning rates.

**Comment by araleza** [comment]
That's interesting, @recris. Back when I was complaining about the vertical stripes, I found I actually had alpha masked most of my images in a way that didn't preserve the RGB color channel for alpha=0 pixels in the resulting .png file. (It's a checkbox on Gimp .png export, but it doesn't always work even when you tick it).

Once I fixed that, my stripe issues became much milder. But they were still around. If the theory is that they're caused by training images that have non 0.0/1.0 mean/std values in their latents, then that would explain why I still get them, but more slowly.

I think I'll try to write that code to print out which training images I have that are producing out-of-tolerance mean/std values in latents this weekend. And I want to give your code snippet a try too. It might even improve training quality with the un-fixed training images.

**Comment by araleza** [comment]
I gave that code a try on a training run that had already started showing vertical lines, to see if it could repair/recover it. No such luck - the lines stay around. And it also started making very 'grainy' images, like film grain, and kinda darker shadows on the images too. I don't think it's working out for me.

I put it after:
```
    if args.masked_loss or ("alpha_masks" in batch and batch["alpha_masks"] is not None):
        loss = apply_masked_loss(loss, batch)
```
and my training images are mostly 95% alpha, apart from the (smallish) object I'm training to learn. I don't know if that affects anything or not.

**Comment by recris** [comment]
You have to be careful with loss masks, the way you inserted the change is not going to work because it will apply `dist_loss` across the whole image, but regular loss will not.

I've actually made other changes to the way loss mask is applied, instead of doing `loss * mask_image` I've changed it to apply directly on the gradient during back-propagation via `register_hook`:

```
noise_pred.register_hook(lambda grad: grad * mask)  # mask has same dimensions as the latent tensor
```
This allows me to more easily play with the loss function and not have to worry if the mask is being incorrectly applied.

**Comment by sdbds** [comment]
https://github.com/kohya-ss/sd-scripts/pull/1541
This is a pr function I created that uses the shift conditions mentioned as possible in the paper to adaptively adjust shift based on resolution...
